src/client/dfuse/inval.c

```diff
@@ -0,0 +1,423 @@
+ *  Use arrays rather than lists for the buckets for faster iteration.
+ *  Reference counting the timeout buckets.
+ *
+ * Locking: The ival_lock is contended, it is accessed several places, however none do any more
+ * than list management.  As inodes might be removed from one list and re-inserted into another
+ * there is a pre subsystem lock rather than per list locks.
+ *  ie_close() which is called from forget and some failure paths in readdir()
+ *  lookup() to move entries to the end of this list.
+ *  de_run() to pull items from the front of the list.
+ *
+ * Wakeup: The invalidation thread is woken up when:
+ *  dfuse is exiting.
+ *  something is added to an empty list.
+ *  after a timeout.
+ * Timeouts are chosen based on the entries still on any list, dfuse will sleep as long as it can
+ * but at least 2 seconds and at most 60.
+ * As this relates to releasing resources there is no additional benefit in finer grained time
+ * control than this.
+ */
+
+/* Represents one timeout value (time).  Maintains a ordered list of dentries that are using
+ * this timeout.
+ */
+struct dfuse_time_entry {
+	d_list_t inode_list;
+	double   time;
+	d_list_t dte_list;
+};
+
+/* Core data structure, maintains a list of struct dfuse_time_entry lists.*/
+struct dfuse_ival {
+	d_list_t             time_entry_list;
+	struct fuse_session *session;
+	bool                 session_dead;
+};
+
+/* The core data from struct dfuse_inode_entry.  No additional inode references are held on inodes
+ * because of there place on invalidate lists, rather inodes are removed from any list on close.
+ * Therefore once a decision is made to evict an inode then a copy of the data is needed as once
+ * the ival_lock is dropped the inode could be freed.  This is not a problem if this happens as the
+ * kernel will simply return ENOENT.
+ */
+struct inode_core {
+	char       name[NAME_MAX + 1];
+	fuse_ino_t parent;
+};
+
+/* Number of dentries to invalidate per iteration. This value affects how long the lock is held,
+ * after the invalidations happen then another iteration will start immediately.  Invalidation of
+ * directories however trigger many forget calls so we want to make use of this where possible so
+ * keep this batch size small.
+ */
+#define EVICT_COUNT 8
+
+static pthread_mutex_t   ival_lock = PTHREAD_MUTEX_INITIALIZER;
+static bool              ival_stop;
+static pthread_t         ival_thread;
+static sem_t             ival_sem;
+static struct dfuse_ival ival_data;
+
+/* Eviction loop, run periodically in it's own thread
+ *
+ * Returns true if there is more work to do.  If false then *sleep_time is set in seconds.
+ */
+static bool
+ival_loop(int *sleep_time)
+{
+	struct dfuse_time_entry *dte;
+	struct inode_core        ic[EVICT_COUNT] = {};
+	int                      idx             = 0;
+	double                   sleep           = (60 * 1) - 1;
+
+	D_MUTEX_LOCK(&ival_lock);
```

Comment:
Is there potential for this loop to take a long time and hold this lock for too long? E.g. in the case of needing to evict millions of entries

Reply:
And on that note, consider an extreme case
1. Evict time set to 3 seconds
2. Entry A has 0.1 seconds remaining, so not evicted
3. Loop continues for X seconds
4. Sleep for 2 seconds
5. Start next loop
6. Entry A expired 2 + X seconds ago

Reply:
L101-104 covers this case, 8 entries are copied onto the stack then the lock is dropped.

Your second comment is covered by L70-71, there is no harm in keeping something longer than required, we already keep something for "X + <something> + <sample time>" intentionally so decreasing the sample time at the cost of CPU usage and lock contention is not desireable.

Reply:
Ah, I missed the `EVICT_COUNT`